dask/dataframe/categorical.py

```diff
@@ -20,16 +21,19 @@ def _categorize_block(df, categories, index):
     categories: dict mapping column name to iterable of categories
     """
     df = df.copy()
+    package_name = df.__class__.__module__.split(".")[0]
     for col, vals in categories.items():
         if is_categorical_dtype(df[col]):
             df[col] = df[col].cat.set_categories(vals)
         else:
-            df[col] = pd.Categorical(df[col], categories=vals, ordered=False)
+            cat_dtype = sys.modules[package_name].CategoricalDtype(categories=vals, ordered=False)
+            df[col] = sys.modules[package_name].Series(df[col], dtype=cat_dtype)
```

Comment:
cuf uses `pandas.CategoricalDtype`, right? How about we do this instead, which IIUC should work for pandas or rapids

```python
df[col] = df[col].astype(cat_dtype)
```

Reply:
@TomAugspurger Looks like `pandas.CategoricalDtype` will not accept `cudf.Series` inputs, which is why I had to change to current categories typecast and setting of categories.

Reply:
This is on pandas / cudf DataFrames, right? This will be a bit slower than my proposed way since you'll have to factorization steps for non-categorical input: one for the original data and one from `set_categories`.

What exception did you see when passing a cudf.Series to Categorical?

Reply:
This was the error:
```python
>>> import cudf
>>> s = cudf.Series(['a','b'], dtype='category')
>>> import pandas as pd
>>> pd.CategoricalDtype(s)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 226, in __init__
    self._finalize(categories, ordered, fastpath=False)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 347, in _finalize
    categories = self.validate_categories(categories, fastpath=fastpath)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 517, in validate_categories
    raise TypeError(msg.format(categories))
TypeError: Parameter 'categories' must be list-like, was 0    a
1    b
dtype: category
Categories (2, object): [a, b]

>>> pd.CategoricalDtype(cudf.Series([1,2,3]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 226, in __init__
    self._finalize(categories, ordered, fastpath=False)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 347, in _finalize
    categories = self.validate_categories(categories, fastpath=fastpath)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 517, in validate_categories
    raise TypeError(msg.format(categories))
TypeError: Parameter 'categories' must be list-like, was 0    1
1    2
2    3
dtype: int64
```

Reply:
What's `pd.api.types.is_list_like(cudf.Series(['a', 'b']))`?

Reply:
It gives `False`
```python
>>> pd.api.types.is_list_like(cudf.Series(['a', 'b']))
False
```

Reply:
Thanks. Is cudf.Series an Iterable? `isinstance(cudf.Series(['a']), collections.abc.Iterable)`?

If not, and if you register it as an iterable with `collections.abc.Iterable.register(cudf.Series)`, does `pd.CategoricalDtype` work?

Reply:
@TomAugspurger we want to avoid iterating over a `cudf.Series`. This causes element by element device to host copies which is a worst case performance scenario.

Reply:
Still catching up on the conversation and work done in here, but I think the core issue is that cuDF cannot use `pandas.CategoricalDtype` as that uses a host object for storing the dictionary whereas we have `cudf.CategoricalDtype` that uses a GPU object for storing the dictionary.

I think we need to go to a bit of a lower level and generalize the `CategoricalDtype` instead of just this function.

Reply:
You're saying that's why you don't implement `__iter__`? (you could implement `__iter__` to do a one-time device -> host copy and then iter over that, but that may have surprising memory usage).

Stepping back, we could ideally write `df[col].astype(categorical_dtype)`. Why are we passing a `cudf.Series` to `CategoricalDtype` in the first place? Can we convert that to an Index in the first place?

Reply:
Passing an index would still have similar non-list-like error which happened in case of `cudf.Series`:
```python
>>> pd.CategoricalDtype(cudf.core.index.as_index(cudf.Series([1,2,3])))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 226, in __init__
    self._finalize(categories, ordered, fastpath=False)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 347, in _finalize
    categories = self.validate_categories(categories, fastpath=fastpath)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py", line 517, in validate_categories
    raise TypeError(msg.format(categories))
TypeError: Parameter 'categories' must be list-like, was Int64Index([1, 2, 3], dtype='int64')
>>> 
```